### ISSUE TITLE
test_js_optimizer_parse_error fails on windows

### ISSUE THREAD

**Issue by kripken** [emscripten-core/emscripten#9057]
E.g. https://logs.chromium.org/logs/emscripten-releases/buildbucket/cr-buildbucket.appspot.com/8907205937125136752/+/steps/Emscripten_testsuite__upstream__other_/0/stdout

We thought this was due to `\r\n` on windows, but #9052 failed to fix that, so it needs further investigation.

**Comment by yeputons** [comment]
It looks like `emcc`'s output mixes Unix and Windows newlines:

```
...
00000080  33 39 29 0d 0a 76 61 72  20 41 53 4d 5f 43 4f 4e  |39)..var ASM_CON|
00000090  53 54 53 20 3d 20 5b 66  75 6e 63 74 69 6f 6e 28  |STS = [function(|
000000a0  29 20 7b 76 61 72 20 78  20 3d 20 21 3c 2d 3e 35  |) {var x = !<->5|
000000b0  2e 3b 7d 5d 3b 0d 0d 0a  20 20 20 20 20 20 20 20  |.;}];...        |
000000c0  20 20 20 20 20 20 20 20  20 20 20 20 20 20 20 20  |                |
000000d0  20 20 20 20 20 20 20 20  20 20 20 20 20 20 20 5e  |               ^|
000000e0  0d 0a 0d 0a 20 20 20 20  61 74 20 50 61 72 73 65  |....    at Parse|
000000f0  72 2e 70 70 24 34 2e 72  61 69 73 65 20 28 43 3a  |r.pp$4.raise (C:|
...
```

Pay special attention to line `000000b0`: it contains `0d 0d 0a`, which is a Unix newline followed by Windows newline.

Hence, whenever #9052 removes all `0a` from the output, it gets two newlines instead of one.

Looks like `emcc`'s issue to me, not tests issue.

**Comment by yeputons** [comment]
Digging a little more: the fix applied in #9052 does not work because the test calls `expect_fail` which calls `run_process` which sets `universal_newlines=True` when calling `Popen`, which makes Python process all `\r`s and `\n` as it sees fit and leave `\n` only. Apparently, Python replaces unmatched `\r` with `\n`, even though Windows terminal simply ignores it (and thus makes output of `emcc` appear normal).

Unfortunately, quick fix by calling `self.expect_fail([PYTHON, EMCC, 'src.cpp', '-O2'], universal_newlines=False)` instead does not work because `expect_fail` delegates `universal_newlines=False` to `run_process` (`shared/tools.py`) which simply ignores it due to #9069.

So, my suggestion is to either fix #9068 directly and remove #9052 (as it does not do anything), or to fix #9069 first and then call `expect_fail` with `universal_newlines=False` while leaving #9052 in place.

**Comment by tlively** [comment]
Thanks for digging into this, @yeputons! Would you like to submit PRs for the problems you identified?

**Comment by kripken** [comment]
Ok, the first PR here landed (#9070) - let's see that the windows CI is green after that.

If so, then I think we still have followup work here to do. I'm not sure which of the options here is best, though?

**Comment by yeputons** [comment]
I think it's the best to fix the underlying issue: #9088
